internal/engine/up_watch.go

```diff
@@ -0,0 +1,254 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+
+	"github.com/windmilleng/fsnotify"
+	"github.com/windmilleng/tilt/internal/git"
+	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/watch"
+)
+
+type serviceFilesChangedEvent struct {
+	service model.Service
+	files   []string
+}
+
+type serviceWatcher struct {
+	events <-chan serviceFilesChangedEvent
+	errs   <-chan error
+}
+
+// returns a serviceWatcher that tells its reader when a service's file dependencies have changed
+func makeServiceWatcher(
+	ctx context.Context,
+	watcherMaker watcherMaker,
+	timerMaker timerMaker,
+	services []model.Service) (*serviceWatcher, error) {
+
+	watchEventsStream, err := makeWatchEventsStream(ctx, watcherMaker, timerMaker, services)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceChan := make(chan serviceFilesChangedEvent)
+	errs := make(chan error)
+	go func() {
+		// a map of service names to a list of any unprocessed changes
+		leftover := make(map[string]*serviceFilesChangedEvent)
```

Comment:
talked about this offline, but i think this layer is unnecessary, and can be mostly deleted without changing behavior